infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,252 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use digest::Digest;
+use tari_crypto::keys::PublicKey;
+
+#[derive(Debug, Error)]
+pub enum NodeIDError {
+    OutOfBounds,
+}
+
+/// Calculate the distance between two node ids using the XOR metric
+pub fn xor_metric(x: &NodeID, y: &NodeID) -> NodeDistance {
+    let mut nd = NodeDistance::new();
+    for i in 0..nd.0.len() {
+        nd.0[i] = x.0[i] ^ y.0[i];
+    }
+    (nd)
+}
+
+#[derive(Clone, Debug, Eq, PartialOrd, Ord)]
+pub struct NodeDistance([u8; 32]);
+
+impl NodeDistance {
+    /// Construct a new zero distance
+    pub fn new() -> NodeDistance {
+        (NodeDistance::from(&[0; 32]))
+    }
+
+    /// Construct a node distance from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeDistance {
+        (NodeDistance(*array))
+    }
+}
+
+impl PartialEq for NodeDistance {
+    fn eq(&self, nd: &NodeDistance) -> bool {
+        for i in 0..self.0.len() {
+            if self.0[i] != nd.0[i] {
+                return false;
+            }
+        }
+        (true)
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct NodeID([u8; 32]);
```

Comment:
Should be `NodeId` as per rust styleguide - https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

> In CamelCase, acronyms count as one word: use Uuid rather than UUID.

Reply:
Change applied